packages/neon_http_client/lib/src/client.dart

```diff
@@ -0,0 +1,113 @@
+}
+
+/// An exception caused by a [HttpInterceptor].
+final class InterceptionFailure extends NeonClientException {
+  /// Creates a new interceptor failure exception.
+  InterceptionFailure(super.message, [super.uri]);
+}
+
+/// A client configuring the clients for all Nextcloud APIs.
+///
+/// To access the APIs of a particular app import the extensions through `package:nextcloud/{id}.dart`.
+final class NeonClient with http.BaseClient {
+  /// Creates a new Neon http client.
+  ///
+  ///
+  /// It is good practice to set the [userAgent] to allow server admins to identify clients.
+  /// A custom HTTP client can be provided through [client].
+  /// Additionally a [cookieStore] can be specified to save cookies across requests.
+  /// Some endpoints require the use of a cookies persistence.
+  NeonClient({
+    http.Client? client,
+    Iterable<HttpInterceptor>? interceptors,
+    String? userAgent,
+    CookieStore? cookieStore,
+  })  : _baseClient = client ?? http.Client(),
+        _interceptors = BuiltList.build((builder) {
+          if (interceptors != null) {
+            builder.addAll(interceptors);
+          }
+
+          if (cookieStore != null) {
+            builder.add(
+              CookieStoreInterceptor(cookieStore: cookieStore),
+            );
+          }
+
+          if (userAgent != null) {
+            builder.add(
+              BaseHeaderInterceptor(
+                baseHeaders: {
+                  HttpHeaders.userAgentHeader: userAgent,
+                },
+              ),
+            );
+          }
+        });
+
+  final http.Client _baseClient;
+
+  final BuiltList<HttpInterceptor> _interceptors;
+
+  @override
+  Future<http.StreamedResponse> send(http.BaseRequest request) async {
+    var interceptedRequest = request;
+    for (final interceptor in _interceptors) {
+      if (interceptor.shouldInterceptRequest(interceptedRequest)) {
+        try {
+          interceptedRequest = await interceptor.interceptRequest(
+            request: interceptedRequest,
+          );
+        } catch (_, stackTrace) {
+          Error.throwWithStackTrace(
+            InterceptionFailure('Failed to intercept request', request.url),
+            stackTrace,
+          );
+        }
+      }
+    }
+
+    var interceptedResponse = await _baseClient.send(interceptedRequest);
+
+    Uri url;
+    if (interceptedResponse case http.BaseResponseWithUrl(url: final responseUrl)) {
+      url = responseUrl;
+    } else {
+      url = interceptedRequest.url;
+    }
+
+    for (final interceptor in _interceptors) {
+      if (interceptor.shouldInterceptResponse(interceptedResponse)) {
+        try {
+          interceptedResponse = await interceptor.interceptResponse(
+            response: interceptedResponse,
+            url: url,
+          );
+        } catch (_, stackTrace) {
+          Error.throwWithStackTrace(
+            InterceptionFailure('Failed to intercept response', request.url),
```

Comment:
```suggestion
            InterceptionFailure('Failed to intercept response', url),
```

Reply:
This is on purpose.
It's easier to debug a thrown exception with the url the request was made against, rather than having to figure out what redirects ended up giving the response.

what do you think?

Reply:
But knowing the final url is also helpful :thinking:
Then just leave it as it was.

Reply:
Yeah I'll leave it.
I don't really expect our interceptors to throw anything meaningful. They can only throw deeper errors, like a sqlite exception from the cookie persistence.
In that case, the url isn't that helpful anyway.